src/tree.ts

```diff
@@ -0,0 +1,560 @@
+  leaked memory for these reasons:
+  1) mathField.revert() didn't actually call the .dispose() method
+     on ANY of the nodes.
+  2) parts of the code create temporary nodes that never get linked
+     to anything. So they definitely didn't get their dispose() method
+     called.
+  3) even if everything above worked it's really common for users of
+     mathquill to forget to tear it down correctly.
+
+  It turns out mathquill always uses the Node and the Element as pairs. So we
+  can store the Node on the Element and the Element on the Node. That makes it
+  possible to get one from the other. This also has the added benefit of meaning
+  the Node isn't stored in a global dictionary. If you lose all references to
+  the Element, then you'll also lose all references to the Node. This means the
+  browser can garbage collect all of mathquill's internals when the DOM is destroyed.
+
+  There's only 1 small gotcha. The linking between Element and Node is a little clumsy.
+  1) All of the Nodes will be created.
+  2) Then all of the Elements will be created.
+  3) Then the two will be linked
+
+  The problem is that the linking step only has access to the elements. It doesn't have
+  access to the nodes. That means we need to store the id of the node we want on the element
+  at creation time. Then we need to lookup that Node by id during the linking step. This
+  means we still need a dictionary. But at least it can be a temporary dictionary.
+  Steps 1 - 3 happen synchronously. So after those steps we can simply clean out the
+  temporary dictionary and remove all hard references to the Nodes.
+
+  Any time we create a Node we schedule a task to clean all Nodes out of the dictionary
+  on the next frame. That's safe because there's no opportunity for nodes to be created
+  and NOT linked between the time we schedule the cleaning step and actually do it.
+*/
+
+function eachNode (ends:Ends, yield_:(el:MQNode) => boolean | undefined) {
+  var el = ends[L];
+  if (!el) return;
+
+  var stop = ends[R];
+  if (!stop) return; //shouldn't happen because ends[L] is defined;
+  stop = stop[R];
+
+  // TODO - this cast as MQNode is actually important to keep tests passing. I went to
+  // fix this code to gracefully handle an undefined el and there are tests that actually
+  // verify that this will throw an error. So I'm keeping the behavior but ignoring the
+  // type error.
+  for (; el !== stop; el = (el as MQNode)[R]) {
+    var result = yield_(el as MQNode); // TODO - might be passing in 0 intead of a MQNode, but tests want this
+    if (result === false) break;
+  }
+}
+
+function foldNodes <T>(ends:Ends, fold:T, yield_:(fold:T, el:MQNode) => T):T {
+  var el = ends[L];
+  if (!el) return fold;
+
+  var stop = ends[R];
+  if (!stop) return fold; //shouldn't happen because ends[L] is defined;
+  stop = stop[R];
+
+  // TODO - this cast as MQNode is actually important to keep tests passing. I went to
```

Comment:
New TODO, would be good to open a ticket

Reply:
Tracked as https://github.com/mathquill/mathquill/issues/986